components/indexer.py

```diff
@@ -1,56 +1,55 @@
-
-    is_red = magicbot.tunable(False)
-
-    indexer_speed = magicbot.tunable(0.8)
-
+class Indexer:
+    indexer_front_motor: rev.CANSparkMax
+    indexer_back_motor: rev.CANSparkMax
     colour_sensor: rev.ColorSensorV3
-    indexer_motor: ctre.TalonSRX
-
-    def setup(self):
-        self.indexer_motor.setInverted(False)
-
-    @magicbot.state(first=True)
-    def indexing(self):
-        self.indexer_motor.set(ctre.ControlMode.PercentOutput, self.indexer_speed)
-        if self.has_ball():
-            self.done()
-
-    @magicbot.default_state
-    def stopped(self):
-        self.indexer_motor.set(ctre.ControlMode.PercentOutput, 0)
-
-    @magicbot.timed_state(duration=1, must_finish=True)
-    def clearing(self):
-        self.indexer_motor.set(ctre.ControlMode.PercentOutput, -self.indexer_speed)
-
-    @magicbot.timed_state(duration=0.5, must_finish=True)
-    def firing(self):
-        self.indexer_motor.set(ctre.ControlMode.PercentOutput, self.indexer_speed)
-
-    @magicbot.feedback
-    def has_ball(self) -> bool:
+    breakbeam_sensor: wpilib.DigitalInput
+    # indexer_piston: wpilib.Solenoid
+
+    is_firing = tunable(False)
+    is_red = tunable(False)
+    indexer_speed = tunable(0.5)
+    # Front, mid, back
+    speeds = (0.0, 0.0)
+
+    def setup(self) -> None:
+        self.indexer_front_motor.setInverted(True)
+        self.indexer_back_motor.setInverted(False)
+
+    def execute(self) -> None:
+        # print(self.speeds)
+        self.indexer_front_motor.set(self.speeds[0])
+        self.indexer_back_motor.set(self.speeds[1])
+
+    @feedback
+    def has_back(self) -> bool:
+        return not self.breakbeam_sensor.get()
+
+    @feedback
+    def has_front(self) -> bool:
         return self.colour_sensor.getProximity() > 400
 
-    def can_read(self) -> bool:
-        return self.colour_sensor.getProximity() > 500
+    @feedback
+    def is_front_ours(self) -> bool:
+        raw = self.colour_sensor.getRawColor()
+        return (raw.red > raw.blue) == self.is_red
+
+    @feedback
+    def get_colours(self) -> str:
+        raw = self.colour_sensor.getRawColor()
+        return f"r{raw.red:.3f}, g{raw.green:.3f}, b{raw.blue:.3f}"
 
-    @magicbot.feedback
-    def is_ball_ours(self) -> bool:
-        """Assumes we have a ball we can read"""
-        values = self.colour_sensor.getRawColor()
-        return (values.red > values.blue) == self.is_red
+    @feedback
+    def get_speeds(self) -> str:
+        return f"{self.speeds[0]}, {self.speeds[1]}"
 
-    @magicbot.feedback
-    def colour_values(self):
-        values = self.colour_sensor.getRawColor()
-        return f"R: {values.red}, G: {values.green}, B: {values.green}"
+    def set(self, front: int, back: int) -> None:
+        self.speeds = tuple(self.indexer_speed * s for s in (front, back))
 
-    @magicbot.feedback
-    def colour_prox(self):
-        return self.colour_sensor.getProximity()
+    def set_piston(self, on: bool) -> None:
```

Comment:
What is the piston actually for? This should describe the real action you want to perform.